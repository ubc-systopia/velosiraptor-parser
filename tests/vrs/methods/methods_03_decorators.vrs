segment MyUnit {
    // method with pre-condition
    #[prop]
    fn foo() -> bool
        requires true;

    // method with post-condition
    #[prop, prop2]
    fn foo() -> bool
        ensures true;

    // method with pre and post conditions
    #[prop(arg)]
    fn foo() -> bool
        requires true;
        ensures true;


    // method with pre-condition
    #[prop(arg), prop2]
    fn foo() -> bool
        requires true;
    {
        true
    }

    // method with an argument
    #[prop(arg), prop2]
    fn bar(val: int) -> int
        requires true;
        requires val > 0 && true;
        ensures bar(val) == val + 1;
        ensures true;
    {
        1 + val
    }
}