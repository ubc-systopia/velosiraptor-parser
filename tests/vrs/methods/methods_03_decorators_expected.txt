VelosiParseTree(tests/vrs/methods/methods_03_decorators.vrs)
---------------------------------------------
segment MyUnit {
  #[prop]
  fn foo() -> bool
    requires true

  #[prop, prop2]
  fn foo() -> bool
    ensures true

  #[prop(arg)]
  fn foo() -> bool
    requires true
    ensures true

  #[prop(arg), prop2]
  fn foo() -> bool
    requires true
  {
    true
  }

  #[prop(arg), prop2]
  fn bar(val: int) -> int
    requires true
    requires (val > 0) && true
    ensures bar(val) == (val + 1)
    ensures true
  {
    1 + val
  }

}


---------------------------------------------
